File: reference-code/backend-v2/services/context_builder.py
"""Context Builder for Smart Chat Agent Context Management.

This module provides intelligent file relevance scoring and context building
for the Chat Agent. It analyzes user queries and scores project files by
relevance, enabling the agent to receive the most useful files upfront
without needing to search.

Key classes:
- RelevanceScorer: Scores files based on keyword matches, file type, and recency
- ContextBuilder: Builds optimized context from scored files (Story 1.2)
- SmartContext: Data structure for context results (Story 1.2)
"""

import re
import time
import logging
from dataclasses import dataclass, field
from typing import Optional

# Configure logger for context builder
logger = logging.getLogger(__name__)


def configure_context_logging(level: int = logging.INFO) -> None:
    """Configure logging for the context builder module.

    Args:
        level: Logging level (e.g., logging.DEBUG, logging.INFO, logging.WARNING)

    Usage:
        from services.context_builder import configure_context_logging
        import logging
        configure_context_logging(logging.DEBUG)  # Enable detailed logs
        configure_context_logging(logging.WARNING)  # Disable most logs
    """
    handler = logging.StreamHandler()
    handler.setFormatter(logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
    ))
    logger.addHandler(handler)
    logger.setLevel(level)


# Configuration constants
MAX_CONTEXT_TOKENS = 4000
MAX_FULL_FILES = 5
MIN_SCORE_THRESHOLD = 0.1

# File type priorities (higher = more likely to be relevant)
FILE_TYPE_PRIORITIES = {
    ".tsx": 1.0,   # React components - most user-facing
    ".ts": 0.8,    # TypeScript utilities, hooks, types
    ".css": 0.6,   # Stylesheets
    ".scss": 0.6,  # SCSS stylesheets
}
DEFAULT_FILE_PRIORITY = 0.4

# Stop words to filter from queries
STOP_WORDS = {"the", "a", "an", "to", "make", "change", "update", "add", "fix", "is", "it", "in", "on", "for"}

# Route-related keywords for page file bonus
ROUTE_KEYWORDS = {"page", "route", "navigate", "navigation", "home", "homepage", "/"}


class RelevanceScorer:
    """Scores files by relevance to a user query.

    Uses multiple signals:
    - Keyword matches in content
    - File type priority (.tsx > .ts > .css)
    - Component name matches
    - Page file bonus for route queries
    - Recency bonus for recently modified files
    """

    def __init__(self, recent_files: Optional[list[str]] = None):
        """Initialize the scorer.

        Args:
            recent_files: List of recently modified file paths, ordered by recency
                         (most recent first). Used for recency bonus.
        """
        self.recent_files = recent_files or []

    def score_file(self, file_path: str, file_content: str, query: str) -> float:
        """Score a file's relevance to a query.

        Args:
            file_path: Path to the file (e.g., "components/Header.tsx")
            file_content: Content of the file
            query: User's query string

        Returns:
            Float score between 0.0 and 1.0
        """
        keywords = self._extract_keywords(query)

        # Calculate individual scores
        keyword_score = self._calculate_keyword_score(file_content, keywords)
        file_type_score = self._get_file_type_priority(file_path)
        component_score = self._calculate_component_match(file_path, query)
        route_score = self._calculate_route_bonus(file_path, query)
        recency_score = self._calculate_recency_bonus(file_path)

        # Weighted combination (from story dev notes)
        final_score = (
            keyword_score * 0.4 +       # 40% weight on keyword matches
            file_type_score * 0.25 +    # 25% weight on file type
            component_score * 0.2 +     # 20% weight on component name match
            route_score * 0.1 +         # 10% weight on page file bonus
            recency_score * 0.05        # 5% weight on recency
        )

        # Clamp to 0.0-1.0 range
        return max(0.0, min(1.0, final_score))

    def _extract_keywords(self, query: str) -> list[str]:
        """Extract meaningful keywords from a query.

        Args:
            query: User's query string

        Returns:
            List of lowercase keywords (>= 3 chars, no stop words)
        """
        # Lowercase and split on whitespace/punctuation
        words = re.split(r'[\s\-_.,;:!?()]+', query.lower())

        # Filter: remove stop words and short words
        keywords = [
            word for word in words
            if len(word) >= 3 and word not in STOP_WORDS
        ]

        return keywords

    def _calculate_keyword_score(self, content: str, keywords: list[str]) -> float:
        """Calculate score based on keyword matches in content.

        Args:
            content: File content to search
            keywords: List of keywords to look for

        Returns:
            Float score between 0.0 and 1.0
        """
        if not keywords:
            return 0.0

        content_lower = content.lower()
        matches = sum(1 for kw in keywords if kw in content_lower)

        # Return ratio of matched keywords (capped at 1.0)
        return min(1.0, matches / len(keywords))

    def _get_file_type_priority(self, file_path: str) -> float:
        """Get priority score based on file extension.

        Args:
            file_path: Path to the file

        Returns:
            Priority score between 0.0 and 1.0
        """
        # Extract extension
        ext_match = re.search(r'\.[a-zA-Z]+$', file_path)
        if not ext_match:
            return DEFAULT_FILE_PRIORITY

        ext = ext_match.group().lower()
        return FILE_TYPE_PRIORITIES.get(ext, DEFAULT_FILE_PRIORITY)

    def _calculate_component_match(self, file_path: str, query: str) -> float:
        """Calculate bonus for component name matching query.

        Args:
            file_path: Path to the file
            query: User's query string

        Returns:
            0.3 if component name found in query, 0.0 otherwise
        """
        # Extract component name from file path (e.g., "Header.tsx" -> "header")
        filename = file_path.split("/")[-1]
        component_name = re.sub(r'\.[a-zA-Z]+$', '', filename).lower()

        # Check if component name appears in query
        if component_name and len(component_name) >= 3 and component_name in query.lower():
            return 0.3

        return 0.0

    def _calculate_route_bonus(self, file_path: str, query: str) -> float:
        """Calculate bonus for page files when query is route-related.

        Args:
            file_path: Path to the file
            query: User's query string

        Returns:
            0.2 if page file and route query, 0.0 otherwise
        """
        # Check if query mentions routes/pages
        query_lower = query.lower()
        is_route_query = any(kw in query_lower for kw in ROUTE_KEYWORDS)

        if not is_route_query:
            return 0.0

        # Check if file is a page file
        filename = file_path.split("/")[-1].lower()
        is_page_file = filename in ("page.tsx", "layout.tsx", "page.ts", "layout.ts")

        return 0.2 if is_page_file else 0.0

    def _calculate_recency_bonus(self, file_path: str) -> float:
        """Calculate bonus based on how recently the file was modified.

        Args:
            file_path: Path to the file

        Returns:
            Score between 0.0 and 0.15 based on recency position
        """
        if not self.recent_files or file_path not in self.recent_files:
            return 0.0

        # Get position (0 = most recent)
        position = self.recent_files.index(file_path)

        # Calculate bonus: most recent gets 0.15, decreasing for older
        max_bonus = 0.15
        decay_factor = 0.7  # Each older position gets 70% of previous bonus

        bonus = max_bonus * (decay_factor ** position)
        return bonus


# =============================================================================
# Data Classes for Context Structure (Story 1.2)
# =============================================================================

@dataclass
class FileContent:
    """Represents a file with its full content included in context.

    Attributes:
        path: File path relative to project root
        content: Full file content
        score: Relevance score (0.0-1.0)
    """
    path: str
    content: str
    score: float


@dataclass
class FileSummary:
    """Represents a file summary for files not included in full.

    Attributes:
        path: File path relative to project root
        line_count: Number of lines in the file
        purpose: Brief description of file purpose
        score: Relevance score (0.0-1.0)
    """
    path: str
    line_count: int
    purpose: str
    score: float


@dataclass
class SmartContext:
    """Container for intelligently built context.

    Attributes:
        full_files: Files included with full content
        summaries: Files included as summaries only
        token_count: Estimated total tokens used
    """
    full_files: list[FileContent] = field(default_factory=list)
    summaries: list[FileSummary] = field(default_factory=list)
    token_count: int = 0


# =============================================================================
# Helper Functions (Story 1.2)
# =============================================================================

def estimate_tokens(text: str) -> int:
    """Estimate the number of tokens in a text string.

    Uses a simple approximation of ~4 characters per token.
    This is conservative to avoid exceeding context limits.

    Args:
        text: The text to estimate tokens for

    Returns:
        Estimated token count
    """
    if not text:
        return 0
    return len(text) // 4


def estimate_file_tokens(file_path: str, content: str) -> int:
    """Estimate tokens for a file including formatting overhead.

    Accounts for the markdown formatting used when including files:
    ### {path}
    ```
    {content}
    ```

    Args:
        file_path: Path to the file
        content: File content

    Returns:
        Estimated token count including formatting
    """
    # Format: "### {path}\n```\n{content}\n```\n"
    formatted = f"### {file_path}\n```\n{content}\n```\n"
    return estimate_tokens(formatted)


def detect_file_purpose(file_path: str, content: str = "") -> str:
    """Detect the purpose of a file based on its path and content.

    Args:
        file_path: Path to the file
        content: File content (optional, for future enhancement)

    Returns:
        Brief purpose string (max 30 chars)
    """
    path_lower = file_path.lower()
    filename = file_path.split("/")[-1].lower()

    # Page/Layout files
    if filename in ("page.tsx", "page.ts"):
        return "Page component"
    if filename in ("layout.tsx", "layout.ts"):
        return "Layout component"

    # By directory
    if "/components/" in path_lower or path_lower.startswith("components/"):
        return "React component"
    if "/hooks/" in path_lower or path_lower.startswith("hooks/"):
        return "React hook"
    if "/lib/" in path_lower or path_lower.startswith("lib/"):
        return "Utility library"
    if "/utils/" in path_lower or path_lower.startswith("utils/"):
        return "Utility functions"
    if "/types/" in path_lower or path_lower.startswith("types/"):
        return "Type definitions"
    if "/styles/" in path_lower or path_lower.startswith("styles/"):
        return "Stylesheet"
    if "/api/" in path_lower or path_lower.startswith("api/"):
        return "API route"
    if "/services/" in path_lower or path_lower.startswith("services/"):
        return "Service module"

    # By extension
    if filename.endswith(".css") or filename.endswith(".scss"):
        return "Stylesheet"
    if filename.endswith(".tsx"):
        return "React component"
    if filename.endswith(".ts"):
        return "TypeScript module"
    if filename.endswith(".json"):
        return "Configuration"
    if filename.endswith(".md"):
        return "Documentation"

    return "Source file"


# =============================================================================
# Context Builder (Story 1.2)
# =============================================================================

class ContextBuilder:
    """Builds optimized context from scored files for the Chat Agent.

    Uses RelevanceScorer to rank files, then selects the most relevant
    files to include in full within token limits. Remaining files are
    included as summaries.
    """

    def __init__(
        self,
        scorer: Optional[RelevanceScorer] = None,
        max_tokens: int = MAX_CONTEXT_TOKENS,
        max_full_files: int = MAX_FULL_FILES,
        min_score_threshold: float = MIN_SCORE_THRESHOLD
    ):
        """Initialize the context builder.

        Args:
            scorer: RelevanceScorer instance (creates default if None)
            max_tokens: Maximum tokens for full file content
            max_full_files: Maximum number of files to include in full
            min_score_threshold: Minimum score to include a file
        """
        self.scorer = scorer or RelevanceScorer()
        self.max_tokens = max_tokens
        self.max_full_files = max_full_files
        self.min_score_threshold = min_score_threshold

    def build_context(
        self,
        generated_files: dict[str, str],
        query: str,
        max_tokens: Optional[int] = None
    ) -> SmartContext:
        """Build optimized context from project files.

        Args:
            generated_files: Dict mapping file paths to content
            query: User's query string
            max_tokens: Override default max tokens (optional)

        Returns:
            SmartContext with full files and summaries
        """
        start_time = time.perf_counter()

        if not generated_files:
            logger.debug("No files to build context from")
            return SmartContext()

        effective_max_tokens = max_tokens or self.max_tokens
        query_preview = query[:50] + "..." if len(query) > 50 else query
        logger.debug(f"Building context for query: '{query_preview}'")

        # 1. Score all files
        scored_files = [
            (path, content, self.scorer.score_file(path, content, query))
            for path, content in generated_files.items()
        ]

        # 2. Sort by score descending
        scored_files.sort(key=lambda x: x[2], reverse=True)

        # Log top scores
        logger.debug(f"Scored {len(scored_files)} files")
        if scored_files:
            top_scores = [(p, f"{s:.3f}") for p, _, s in scored_files[:5]]
            logger.debug(f"Top 5 scores: {top_scores}")

        # 3. Build context within budget
        full_files: list[FileContent] = []
        summaries: list[FileSummary] = []
        token_count = 0

        for path, content, score in scored_files:
            # Skip files below threshold
            if score < self.min_score_threshold:
                continue

            file_tokens = estimate_file_tokens(path, content)

            # Check if we can include full content
            if (token_count + file_tokens <= effective_max_tokens and
                    len(full_files) < self.max_full_files):
                full_files.append(FileContent(
                    path=path,
                    content=content,
                    score=score
                ))
                token_count += file_tokens
                logger.debug(f"Full file: {path} (score={score:.3f}, tokens={file_tokens})")
            else:
                # Add as summary
                summaries.append(FileSummary(
                    path=path,
                    line_count=content.count('\n') + 1,
                    purpose=detect_file_purpose(path, content),
                    score=score
                ))

        # Log summary
        duration_ms = (time.perf_counter() - start_time) * 1000
        logger.info(
            f"Context built: {len(full_files)} full files, "
            f"{len(summaries)} summaries, "
            f"{token_count} tokens, "
            f"{duration_ms:.1f}ms"
        )

        return SmartContext(
            full_files=full_files,
            summaries=summaries,
            token_count=token_count
        )

File: lib/agents/context-builder.ts
/**
 * Smart Context Builder
 *
 * Builds optimized context for the Chat Agent by scoring files by relevance
 * to the user's query. This reduces tool calls by ~75% by pre-loading
 * the most relevant files.
 *
 * Ported from: reference-code/backend-v2/services/context_builder.py
 */

import type {
  SmartContext,
  SmartContextConfig,
  FileContent,
  FileSummary,
} from "./types";

/**
 * Request complexity levels (local definition to avoid circular dependency)
 * @deprecated Legacy complexity enum — kept for backward compatibility
 */
export enum RequestComplexity {
  SIMPLE = "simple",
  MODERATE = "moderate",
  COMPLEX = "complex",
  ARCHITECTURAL = "architectural",
}

/**
 * Dynamic context configuration based on request complexity
 */
export interface DynamicContextConfig extends SmartContextConfig {
  /** Maximum number of files to include */
  maxFiles: number;
  /** Recursion limit for agent */
  recursionLimit: number;
}

/**
 * Get context configuration based on request complexity
 */
export function getContextConfigForComplexity(
  complexity: RequestComplexity
): DynamicContextConfig {
  switch (complexity) {
    case RequestComplexity.SIMPLE:
      return {
        maxTokens: 3000,
        maxFullFiles: 3,
        minScoreThreshold: 0.15,
        maxFiles: 3,
        recursionLimit: 30,
      };
    case RequestComplexity.MODERATE:
      return {
        maxTokens: 5000,
        maxFullFiles: 6,
        minScoreThreshold: 0.1,
        maxFiles: 6,
        recursionLimit: 50,
      };
    case RequestComplexity.COMPLEX:
      return {
        maxTokens: 8000,
        maxFullFiles: 10,
        minScoreThreshold: 0.08,
        maxFiles: 10,
        recursionLimit: 150,
      };
    case RequestComplexity.ARCHITECTURAL:
      return {
        maxTokens: 10000,
        maxFullFiles: 15,
        minScoreThreshold: 0.05,
        maxFiles: 15,
        recursionLimit: 200,
      };
    default:
      return {
        maxTokens: 4000,
        maxFullFiles: 5,
        minScoreThreshold: 0.1,
        maxFiles: 5,
        recursionLimit: 50,
      };
  }
}

// Configuration constants
const DEFAULT_MAX_CONTEXT_TOKENS = 4000;
const DEFAULT_MAX_FULL_FILES = 5;
const DEFAULT_MIN_SCORE_THRESHOLD = 0.1;

// File type priorities (higher = more likely to be relevant)
const FILE_TYPE_PRIORITIES: Record<string, number> = {
  ".tsx": 1.0, // React components - most user-facing
  ".ts": 0.8, // TypeScript utilities, hooks, types
  ".css": 0.6, // Stylesheets
  ".scss": 0.6, // SCSS stylesheets
};
const DEFAULT_FILE_PRIORITY = 0.4;

// Stop words to filter from queries (excluding action verbs)
const STOP_WORDS = new Set([
  "the",
  "a",
  "an",
  "to",
  "is",
  "it",
  "in",
  "on",
  "for",
  "of",
  "and",
  "or",
  "but",
  "with",
  "this",
  "that",
  "be",
  "are",
  "was",
  "were",
  "been",
  "being",
  "have",
  "has",
  "had",
  "do",
  "does",
  "did",
  "will",
  "would",
  "could",
  "should",
  "may",
  "might",
  "must",
  "can",
  "i",
  "you",
  "we",
  "they",
  "please",
  "want",
  "need",
  "like",
]);

// Critical action verbs that should NEVER be filtered
const ACTION_VERBS = new Set([
  "add",
  "fix",
  "change",
  "make",
  "update",
  "remove",
  "delete",
  "create",
  "modify",
  "edit",
  "replace",
  "move",
  "rename",
  "copy",
  "insert",
  "append",
  "prepend",
  "hide",
  "show",
  "enable",
  "disable",
  "implement",
  "refactor",
  "optimize",
  "improve",
]);

// Route-related keywords for page file bonus
const ROUTE_KEYWORDS = new Set([
  "page",
  "route",
  "navigate",
  "navigation",
  "home",
  "homepage",
  "/",
]);

// Semantic groups for concept-based matching
const SEMANTIC_GROUPS: Record<string, string[]> = {
  navigation: ["nav", "navbar", "header", "menu", "sidebar", "navigation", "link", "links"],
  styling: ["style", "css", "color", "theme", "design", "appearance", "look", "visual"],
  button: ["button", "btn", "click", "action", "cta", "submit"],
  form: ["form", "input", "field", "submit", "validation", "text", "textarea"],
  layout: ["layout", "grid", "flex", "container", "wrapper", "section", "spacing"],
  card: ["card", "tile", "box", "panel", "widget"],
  modal: ["modal", "dialog", "popup", "overlay", "drawer"],
  list: ["list", "table", "grid", "items", "collection", "array"],
  auth: ["auth", "login", "signup", "register", "password", "user", "account"],
  data: ["data", "api", "fetch", "load", "state", "store", "context"],
  error: ["error", "warning", "alert", "notification", "toast", "message"],
  image: ["image", "img", "photo", "picture", "icon", "logo", "avatar"],
  text: ["text", "heading", "title", "paragraph", "label", "copy", "content"],
  animation: ["animation", "transition", "motion", "hover", "animate", "effect"],
};

/**
 * Relevance Scorer
 *
 * Scores files by relevance to a user query using multiple signals.
 */
export class RelevanceScorer {
  private recentFiles: string[];

  constructor(recentFiles: string[] = []) {
    this.recentFiles = recentFiles;
  }

  /**
   * Score a file's relevance to a query.
   *
   * @param filePath - Path to the file
   * @param fileContent - Content of the file
   * @param query - User's query string
   * @returns Float score between 0.0 and 1.0
   */
  scoreFile(filePath: string, fileContent: string, query: string): number {
    const keywords = this.extractKeywords(query);

    // Calculate individual scores
    const keywordScore = this.calculateKeywordScore(fileContent, keywords);
    const fileTypeScore = this.getFileTypePriority(filePath);
    const componentScore = this.calculateComponentMatch(filePath, query);
    const routeScore = this.calculateRouteBonus(filePath, query);
    const recencyScore = this.calculateRecencyBonus(filePath);
    const semanticScore = this.calculateSemanticBonus(filePath, fileContent, keywords);

    // Weighted combination (adjusted to include semantic matching)
    const finalScore =
      keywordScore * 0.35 + // 35% weight on keyword matches (reduced from 40%)
      fileTypeScore * 0.2 + // 20% weight on file type (reduced from 25%)
      componentScore * 0.2 + // 20% weight on component name match
      routeScore * 0.1 + // 10% weight on page file bonus
      recencyScore * 0.05 + // 5% weight on recency
      semanticScore * 0.1; // 10% weight on semantic matching (NEW)

    // Clamp to 0.0-1.0 range
    return Math.max(0.0, Math.min(1.0, finalScore));
  }

  /**
   * Extract meaningful keywords from a query.
   * Preserves action verbs even if they appear in stop words.
   */
  private extractKeywords(query: string): string[] {
    // Lowercase and split on whitespace/punctuation
    const words = query.toLowerCase().split(/[\s\-_.,;:!?()+]+/);

    // Filter: remove stop words and short words, but ALWAYS keep action verbs
    return words.filter(
      (word) =>
        word.length >= 3 && (ACTION_VERBS.has(word) || !STOP_WORDS.has(word))
    );
  }

  /**
   * Calculate score based on keyword matches in content.
   */
  private calculateKeywordScore(
    content: string,
    keywords: string[]
  ): number {
    if (keywords.length === 0) {
      return 0.0;
    }

    const contentLower = content.toLowerCase();
    const matches = keywords.filter((kw) => contentLower.includes(kw)).length;

    // Return ratio of matched keywords (capped at 1.0)
    return Math.min(1.0, matches / keywords.length);
  }

  /**
   * Get priority score based on file extension.
   */
  private getFileTypePriority(filePath: string): number {
    const extMatch = filePath.match(/\.[a-zA-Z]+$/);
    if (!extMatch) {
      return DEFAULT_FILE_PRIORITY;
    }

    const ext = extMatch[0].toLowerCase();
    return FILE_TYPE_PRIORITIES[ext] ?? DEFAULT_FILE_PRIORITY;
  }

  /**
   * Calculate bonus for component name matching query.
   */
  private calculateComponentMatch(filePath: string, query: string): number {
    // Extract component name from file path
    const filename = filePath.split("/").pop() || "";
    const componentName = filename.replace(/\.[a-zA-Z]+$/, "").toLowerCase();

    // Check if component name appears in query
    if (
      componentName &&
      componentName.length >= 3 &&
      query.toLowerCase().includes(componentName)
    ) {
      return 0.3;
    }

    return 0.0;
  }

  /**
   * Calculate bonus for page files when query is route-related.
   */
  private calculateRouteBonus(filePath: string, query: string): number {
    const queryLower = query.toLowerCase();
    const isRouteQuery = Array.from(ROUTE_KEYWORDS).some((kw) =>
      queryLower.includes(kw)
    );

    if (!isRouteQuery) {
      return 0.0;
    }

    // Check if file is a page file
    const filename = (filePath.split("/").pop() || "").toLowerCase();
    const isPageFile = ["page.tsx", "layout.tsx", "page.ts", "layout.ts"].includes(
      filename
    );

    return isPageFile ? 0.2 : 0.0;
  }

  /**
   * Calculate bonus based on how recently the file was modified.
   */
  private calculateRecencyBonus(filePath: string): number {
    if (
      this.recentFiles.length === 0 ||
      !this.recentFiles.includes(filePath)
    ) {
      return 0.0;
    }

    const position = this.recentFiles.indexOf(filePath);
    const maxBonus = 0.15;
    const decayFactor = 0.7;

    return maxBonus * Math.pow(decayFactor, position);
  }

  /**
   * Calculate semantic similarity bonus based on concept groups.
   * Boosts files that match semantic concepts in the query.
   */
  private calculateSemanticBonus(
    filePath: string,
    fileContent: string,
    keywords: string[]
  ): number {
    const pathLower = filePath.toLowerCase();
    const contentLower = fileContent.toLowerCase();
    let totalBonus = 0;

    // Check each semantic group
    for (const [, synonyms] of Object.entries(SEMANTIC_GROUPS)) {
      // Check if any keyword matches this semantic group
      const keywordMatches = keywords.some((kw) =>
        synonyms.some((syn) => kw.includes(syn) || syn.includes(kw))
      );

      if (!keywordMatches) {
        continue;
      }

      // Check if file path or content contains related concepts
      const fileMatches = synonyms.some(
        (syn) => pathLower.includes(syn) || contentLower.includes(syn)
      );

      if (fileMatches) {
        totalBonus += 0.15;
      }
    }

    // Cap the bonus at 0.3 (matching ~2 semantic groups)
    return Math.min(0.3, totalBonus);
  }
}

/**
 * Estimate the number of tokens in a text string.
 * Uses a simple approximation of ~4 characters per token.
 */
export function estimateTokens(text: string): number {
  if (!text) {
    return 0;
  }
  return Math.ceil(text.length / 4);
}

/**
 * Estimate tokens for a file including formatting overhead.
 */
function estimateFileTokens(filePath: string, content: string): number {
  const formatted = `### ${filePath}\n\`\`\`\n${content}\n\`\`\`\n`;
  return estimateTokens(formatted);
}

/**
 * Detect the purpose of a file based on its path.
 */
export function detectFilePurpose(filePath: string): string {
  const pathLower = filePath.toLowerCase();
  const filename = (filePath.split("/").pop() || "").toLowerCase();

  // Page/Layout files
  if (["page.tsx", "page.ts"].includes(filename)) {
    return "Page component";
  }
  if (["layout.tsx", "layout.ts"].includes(filename)) {
    return "Layout component";
  }

  // By directory
  if (pathLower.includes("/components/") || pathLower.startsWith("components/")) {
    return "React component";
  }
  if (pathLower.includes("/hooks/") || pathLower.startsWith("hooks/")) {
    return "React hook";
  }
  if (pathLower.includes("/lib/") || pathLower.startsWith("lib/")) {
    return "Utility library";
  }
  if (pathLower.includes("/utils/") || pathLower.startsWith("utils/")) {
    return "Utility functions";
  }
  if (pathLower.includes("/types/") || pathLower.startsWith("types/")) {
    return "Type definitions";
  }
  if (pathLower.includes("/styles/") || pathLower.startsWith("styles/")) {
    return "Stylesheet";
  }
  if (pathLower.includes("/api/") || pathLower.startsWith("api/")) {
    return "API route";
  }

  // By extension
  if (filename.endsWith(".css") || filename.endsWith(".scss")) {
    return "Stylesheet";
  }
  if (filename.endsWith(".tsx")) {
    return "React component";
  }
  if (filename.endsWith(".ts")) {
    return "TypeScript module";
  }
  if (filename.endsWith(".json")) {
    return "Configuration";
  }
  if (filename.endsWith(".md")) {
    return "Documentation";
  }

  return "Source file";
}

/**
 * Context Builder
 *
 * Builds optimized context from scored files for the Chat Agent.
 */
export class ContextBuilder {
  private scorer: RelevanceScorer;
  private maxTokens: number;
  private maxFullFiles: number;
  private minScoreThreshold: number;

  constructor(config: SmartContextConfig = {}) {
    this.scorer = new RelevanceScorer();
    this.maxTokens = config.maxTokens ?? DEFAULT_MAX_CONTEXT_TOKENS;
    this.maxFullFiles = config.maxFullFiles ?? DEFAULT_MAX_FULL_FILES;
    this.minScoreThreshold =
      config.minScoreThreshold ?? DEFAULT_MIN_SCORE_THRESHOLD;
  }

  /**
   * Set recent files for recency scoring.
   */
  setRecentFiles(recentFiles: string[]): void {
    this.scorer = new RelevanceScorer(recentFiles);
  }

  /**
   * Build optimized context from project files.
   *
   * @param generatedFiles - Dict mapping file paths to content
   * @param query - User's query string
   * @returns SmartContext with full files and summaries
   */
  buildContext(
    generatedFiles: Map<string, string>,
    query: string
  ): SmartContext {
    if (generatedFiles.size === 0) {
      return {
        fullFiles: [],
        summaries: [],
        tokenCount: 0,
      };
    }

    // 1. Score all files
    const scoredFiles: Array<[string, string, number]> = [];
    for (const [path, content] of generatedFiles.entries()) {
      const score = this.scorer.scoreFile(path, content, query);
      scoredFiles.push([path, content, score]);
    }

    // 2. Sort by score descending
    scoredFiles.sort((a, b) => b[2] - a[2]);

    // 3. Build context within budget
    const fullFiles: FileContent[] = [];
    const summaries: FileSummary[] = [];
    let tokenCount = 0;

    for (const [path, content, score] of scoredFiles) {
      // Skip files below threshold
      if (score < this.minScoreThreshold) {
        continue;
      }

      const fileTokens = estimateFileTokens(path, content);

      // Check if we can include full content
      if (
        tokenCount + fileTokens <= this.maxTokens &&
        fullFiles.length < this.maxFullFiles
      ) {
        fullFiles.push({
          path,
          content,
          score,
        });
        tokenCount += fileTokens;
      } else {
        // Add as summary
        summaries.push({
          path,
          lineCount: content.split("\n").length,
          purpose: detectFilePurpose(path),
          score,
        });
      }
    }

    return {
      fullFiles,
      summaries,
      tokenCount,
    };
  }

  /**
   * Format smart context as a string for the Chat Agent prompt.
   */
  formatContext(context: SmartContext): string {
    const parts: string[] = [];

    // Full files section
    if (context.fullFiles.length > 0) {
      parts.push("## Pre-loaded Files (use directly, no need to read_file)\n");
      for (const file of context.fullFiles) {
        parts.push(`### ${file.path}`);
        parts.push("```");
        parts.push(file.content);
        parts.push("```\n");
      }
    }

    // Summaries section
    if (context.summaries.length > 0) {
      parts.push("## Other Files (use read_file if needed)\n");
      for (const summary of context.summaries) {
        parts.push(
          `- **${summary.path}** (${summary.lineCount} lines) - ${summary.purpose}`
        );
      }
      parts.push("");
    }

    return parts.join("\n");
  }
}
